src/java/org/apache/cassandra/index/SecondaryIndexManager.java

```diff
@@ -362,13 +365,24 @@ public synchronized Future<?> addIndex(IndexMetadata indexDef, boolean isNewCF)
      * @param queryPlan a query plan
      * @throws IndexNotAvailableException if the query plan has any index that is not queryable
      */
-    public void checkQueryability(Index.QueryPlan queryPlan)
+    public boolean isQueryableThroughIndex(Index.QueryPlan queryPlan, boolean allowsFiltering)
```

Comment:
This method can now return true or false, or throw an exception, and it also throws a client warning. That looks like too many side effects for a method starting with `boolean is...`, which might suggest a simpler behaviour. I would either:
a) Split it into two simpler separate boolean methods to know if all the indexes in the plan are building/queryable, and let `ReadCommand#executeLocally` do the AF check and throw exceptions and warnings.
b) Transform it into a `SecondaryIndexManger#searcherFor(Index.QueryPlan, boolean)` method keeping most of it's responsibilities, returning the searcher if it's possible to build it, null if it's building, and exception if it's not queryable.